spikeinterface/sortingcomponents/dimensionality_reduction.py

```diff
@@ -0,0 +1,141 @@
+        n_components : int
+            The number of components to use for the PCA model.
+        detect_peaks_params : dict
+            The parameters for peak detection.
+        peak_selection_params : dict
+            The parameters for peak selection.
+        whiten : bool, optional
+            Whether to whiten the data, by default True.
+
+        {}
+        
+        Returns
+        -------
+        IncrementalPCA: An incremental PCA estimator from scikit-learn.
+            The pca model
+        """
+        
+        # Detect peaks and sub-sample them
+        peaks = detect_peaks(recording, **detect_peaks_params, **job_kwargs)
+        peaks = select_peaks(peaks, **peak_selection_params) # How to select n_peaks
+
+        # Create a waveform extractor
+        ms_before = self._kwargs["ms_before"]
+        ms_after = self._kwargs["ms_after"]
+        # Creates a numpy sorting object where the spike times are the peak times and the unit ids are the peak channel
+        sorting = NumpySorting.from_peaks(peaks, sampling_frequency=recording.sampling_frequency) 
+        we = extract_waveforms(recording, sorting, ms_before=ms_before, ms_after=ms_after, folder=None, 
+                               mode="memory", max_spikes_per_unit=None, **job_kwargs)
+
+        # compute PCA by_channel_global (with sparsity)
+        sparsity = ChannelSparsity.from_radius(we, radius_um=self.local_radius_um)
+        pc = compute_principal_components(we, n_components=n_components,  mode='by_channel_global',
+                                        sparsity=sparsity, whiten=whiten)
+        self.pca_model  = pc.get_pca_model()
+
+        # model_folder should be an input
+        pca_model_path = self._kwargs["model_path"]
+
+        with open(pca_model_path, "wb") as f:
+            pickle.dump(self.pca_model, f)
+            
+        return self.pca_model
+    
+    def compute_buffer(self, traces, peaks, waveforms):        
+
+        if self.pca_model == None:
+            exception_string = (
+                f"Pca model not found, "
+                f"Train the pca model before running compute_buffer using {self.__name__}.fit(kwargs)"
+            )
+            raise AttributeError(exception_string)
+        
+        n_waveforms, n_samples , n_channels = waveforms.shape
+        n_components = self.pca_model.n_components
+        
+        # Initializing non-valid channels with nan (sparsity is used to indicate valid channels)
+        projected_waveforms = np.full(shape =(n_waveforms, n_components, n_channels), fill_value=np.nan)
+
+        for waveform_index, main_channel in enumerate(peaks['channel_ind']):
+            channel_indexes_sparsified,  = np.nonzero(self.neighbours_mask[main_channel])
+            sparsified_waveform = waveforms[waveform_index, :, channel_indexes_sparsified]
+            projected_waveforms[waveform_index, :, channel_indexes_sparsified] = self.pca_model.transform(sparsified_waveform)
```

Comment:
here there is something I do not understand. Sorry.
With the `mode='by_channel_global'` the pca_model is supposed to be applied channel per channel.
Here this is not the case. You transform to a buffer which is (sample, channel). This don't make sens to me.
I need more explanation. Sorry.

Reply:
I think @samuelgarcia  is right. We probably need `sparsified_waveforms.T`, so we input a `(num_channels, num_samples)` array

Reply:
The model is trained to transform samples with `n_timestamps` features (where `n_waveforms, n_timestamps, n_channels = all_waveforms.shape`). What matter for a scikit-learn model is the number of features, the first dimension is the number of samples.

In the code that you have in the principal component module:
```python
        elif mode == "by_channel_global":
            shape = (new_waveforms.shape[0], p['n_components'], num_channels)
            projections = np.zeros(shape)
            for wf_ind, chan_ind in enumerate(channel_inds):
                projections[:, :, chan_ind] = pca_model.transform(new_waveforms[:, :, wf_ind])
```
You extract the samples channel by channel so you end up with a vector of size  (`n_waveforms`, `n_timestamps`), that is, `n_waveforms` samples and `n_timestamps` as features. This works. 

In the code here I extract the samples waveform by waveform so I end up with a vector of size (`n_sparse_channels`, `n_timestamps`, that is, `n_sparse_channels` as samples and `n_timestamps` as features so  it works as well.

I decide to go for the second choice because that way you don't need to load large samples in memory. That is, I did not profile this, maybe the other way is better...